### PR TITLE
feat(aliases): Mangle aliases so they do not clash with column names

### PR DESCRIPTION
**Pull request by lynnagara**
Prefix aliases with `_snuba_`. This was primarily an issue with selected
columns which were assigned aliases that matched the underlying column
names, and prevented us from accessing the actual columns in query processors
where we may want to operate on the raw column.

### CODE REVIEW

**Comment by fpacifici** [comment]
Did you try to run sentry tests against this PR ?

File: snuba/query/parser/__init__.py
Comment (fpacifici) [inline snuba/query/parser/__init__.py:397]: Not following this. What do you mean by "we cannot reference aliases there"

File: snuba/query/parser/__init__.py
Comment (fpacifici) [inline snuba/query/parser/__init__.py:387]: You do not need to rebuild expressions where the alias is not present. You can simply transform the alias, and execute the replace only if it changes (that saves allocation since this happens on the full query).

File: snuba/query/parser/__init__.py
Comment (fpacifici) [inline snuba/query/parser/__init__.py:390]: Why this loop if you already execute the transformation at line 386 ?

File: snuba/query/parser/__init__.py
Comment (fpacifici) [inline snuba/query/parser/__init__.py:392]: Expressions are not transformed in place. This method will return a new expression object that is being ignored here.

File: snuba/query/parser/__init__.py
Comment (lynnagara) [reply snuba/query/parser/__init__.py:389]: I mean that ClickHouse will not run `SELECT exception_stacks.type AS alias FROM sentry_local ARRAY JOIN alias`, so I've rewritten into `SELECT exception_stacks.type AS alias FROM sentry_local ARRAY JOIN exception_stacks.type`

**Review by evanh — COMMENTED** [review]
LGTM as long as Filippo is good with it too.

File: snuba/query/parser/__init__.py
Comment (fpacifici) [reply snuba/query/parser/__init__.py:389]: This works though.

```
SELECT 
    event_id, 
    aaa
FROM sentry_local
ARRAY JOIN tags.key AS aaa

┌─event_id─────────────────────────┬─aaa────┐
│ 855f59b850844461b48349c769470b42 │ color  │
│ 855f59b850844461b48349c769470b42 │ level  │
│ 855f59b850844461b48349c769470b42 │ paying │
└──────────────────────────────────┴────────┘
┌─event_id─────────────────────────┬─aaa────┐
│ 17315b86f5af4859b30e46c2262b993b │ color  │
│ 17315b86f5af4859b30e46c2262b993b │ level  │
│ 17315b86f5af4859b30e46c2262b993b │ paying │
└──────────────────────────────────┴────────┘

6 rows in set. Elapsed: 0.006 sec. 

be29e23e76c8 :) 
```

I think we have three issues here:
1) We could be transforming the column we are using for the `ARRAY JOIN` in several places in the query, but the formatter (which actually assigns aliases) does not care of formatting the ARRAY JOIN clause before everything else so it won't depend on an alias. This is independent from your PR. Could you track this in the backlog ?
2) We are allowing any expression in the AST to be used in the ARRAY JOIN clause. Not a big deal since the language only accepts columns there.
3) In the previous system we were applying aliases to all columns (the alias was the same string as the column name, which were removed when possible during formatting). This is not needed anymore if we do not need to preserve aliases that the client did not provide and since we expand all of them. I think you should not be mangle column aliases when they are the same as the name. You should be able to just remove them. Which basically means undoing what `_apply_column_aliases` does, I don't think you can just skip that step since alias expanding depends on it, though you should be able to remove those that are not needed. This would generate less verbose queries (with fewer chances to fail for their size) and would keep ARRAY JOIN consistent with rest of the query.

**Review by fpacifici — COMMENTED** [review]
I think avoid the mangling on columns where the previous alias was the same as the column name would be a worth a try (shorter queries and no special case for array join). Not 100% it will work though.

**Comment by lynnagara** [comment]
I'm not sure what you mean by avoiding the mangling when alias is the same as column name - isn't the purpose of this PR to do the opposite and add it when that is the case to avoid shadowing?

For example what we are doing now is:
`replaceAll(toString(event_id), '-', '') as event_id`

And we want to change this to:
`replaceAll(toString(event_id), '-', '') AS _snuba_event_id`

And NOT the bare:
`replaceAll(toString(event_id), '-', '')`

Does that line up with your thinking?

Dropping aliases entirely may be a possibility that I haven't thought too much about but I believe it's a much larger change since there no guarantee that our columns at this point represent actual physical ClickHouse columns; rather these "columns" are often mapped to something else entirely in entity and storage processing.

**Comment by fpacifici** [comment]
The bare `replaceAll(toString(event_id), '-', '')` is totally ok. 
The only one that is not ok is `replaceAll(toString(event_id), '-', '') as event_id`

> Dropping aliases entirely may be a possibility that I haven't thought too much about but I believe it's a much larger change since there no guarantee that our columns at this point represent actual physical ClickHouse columns; rather these "columns" are often mapped to something else entirely in entity and storage processing.

In fact you are right. https://github.com/getsentry/snuba/pull/1458 was meant to move in the direction where we would not need to add those aliases early on. But we would still not be ready for that because we still need too doo the reverse mapping from selected aliases to the names provided by the user.
Ignore my comment.

**Review by fpacifici — APPROVED** [review]
Please ensure that you run Sentry tests before merging this and that everything passes.

**Pull request merged**
